### ISSUE TITLE
Incorrect integer scaling after applying Blargg NTSC filters on Miyoo Mini

### ISSUE THREAD

**Issue by mtaras0** [libretro/RetroArch#16041]
## Description

Blargg NTSC filter produces an image of the size [original-height]x602 applied to the standard SNES output, which is double width plus a bit more, due to how it works. It then relies on Retroarch to use integer scaling to scale the image vertically to produce the expected [2x-original-height]x602 image. But when internally calculating integer scaling options Retroarch (I assume) relies on the output image target aspect ratio being exactly 4:3, which results in incorrect vertical scaling with the output resolution of 451x602, no matter the vertical resolution of the actual filter output. This absolutely messes up the image, makes it vertically blurry, messes up with scanlines overlays, etc.

### Expected behavior

Blargg NTSC filter's [original-height]x602 output being scaled correctly to [2x-original-height]x602

### Actual behavior

Output height is instead 451, which is, I assume, is a result of an internal calculation relying on an expected aspect ratio instead of just doubling the existing height, as 451:602 is exactly 4:3

### Steps to reproduce the bug

1. Apply Blargg NTSC filter to SNES output, either as a separate video filter or as a built-in core option
2. Turn on integer scaling, turn off keep "keep image aspect"

### Version/Commit

version 1.15.0, Build date Oct 18 2023

### Environment information

- OS: Linux

**Comment by zoltanvb** [comment]
Does this platform use sdl_dingux graphics driver? There are some hardcoded special values, maybe new ones need to be added:
https://github.com/libretro/RetroArch/blob/aa1d47ad0f04aeac70da5f563b08244605498c5e/gfx/drivers/sdl_dingux_gfx.c#L543

**Comment by mtaras0** [comment]
> Does this platform use sdl_dingux graphics driver?
Yep, it is an sdl_dingux driver. I, however, lack an understanding if specifically what you are linking is what needs to be changed. I can see though that hardcoded values are applied `if (vid->keep_aspect && !vid->integer_scaling)`, but my problems happen specifically with integer scaling On and keep aspect ratio Off (I might just be misunderstanding the linked function though)

**Comment by zoltanvb** [comment]
Existing options are indeed for the case when there is no integer scaling, still I wonder if it could be related, since the actual aspect ratio is never given to this driver, and integer scale / keep aspect is just passed on to firmware (as I understand from the code).
How did you measure the resolution? As I checked now, RA statistics (not sure if it is available at all with this driver) do not reflect if video filter is enabled in RA, and the sanitization should bring the width to 608 instead of 602.
Also, do you see any difference if keep_aspect is enabled or disabled? According to another comment in dingux_util.c, integer_scale implicitly enables keep_aspect.

**Comment by mtaras0** [comment]
I mesured resolution just by screenshots, I sadly didn't find a way to display video statistics on my device (built-in notification system does work for displaying fps, but not for more complex stuff), so yeah, can't chack if 602 gets rounded up to 608.
But the "keep aspect" thing I suspect is important here. I can independently enable and disable "keep aspect" with "integer scaling" turned on, and it gives different results -- when both are enabled the image on the screen displays as a stretched out [original_height]x602 (again just checked via screenshots, so no idea what's actually happening, but visually the image as I expect it to be, which can be different depending on the amoung of cut overscan), which is the correct raw output of the filter, the screen is 640x480 and there's no space to resize this image. With "keep aspect" turned off this stretching to 451 pixels happens.

**Comment by zoltanvb** [comment]
I have not found any nice way to solve this. I suspect that firmware recognizes that aspect ratio is way off, and falls back to non-integer scale, but there is no current logic in this driver to do software scaling.
Maybe we could try and see if there is any sanitized width/height combination that would look acceptable (I do not have the device). Are you able to build RA?

**Comment by mtaras0** [comment]
Sadly I have no experience compiling bigger projects, but I might get back to it later.

For now if anyone stumbles upon this thread I'll say there are two possible fixes:

- The core might support applying Blargg filter by itself, without relying on RA video filters. If this is the case -- apply Blargg filter as a core option, and then use Normal 2x Height video filter to stretch the image correctly vertically (and turn on integer scaling to prevent further scaling)
- The core also might support custom overscan cropping, and we can utilise it: set the image to be cropped to 225 or 226 pixels instead of the standard 224, and set scaling smoothing to "nearest neighbour". This will leave just one line of unimportant pixels on the bottom scaled incorrectly, and everything else is perfectly 2x.

**Comment by mtaras0** [comment]
I've been trying out other systems, and figured out that this problem is not Blargg-filter specific, the integer scaling without "keep aspect ratio" in general seems to be broken. 

For example Game Gear outputs 160x144 image, which is common to scale unevenly since the original screen physically was 4:3 with non square pixels. Integer scaling without keeping aspect ratio should result in a 640x432 image -- 4x horizontally, 3x vertically. But it actually produces 432x576 image -- a 4:3 again.

What seems to be happening is "keep aspect ratio OFF" is being treated as "keep aspect ratio to 4:3 because the screen is 4:3", or something to this effect. In both the Blargg and Game Gear examples the image ends up being 4:3, but in both examples only one dimension is actually integer scaled: Blargg example scales to 1x horizontally (since the filter itself gets the image very wide), but 2.0134x vertically instead of the expected 2x; Game Gear scales to 3x vertically, but to 3.6x horizontally instead of the expected 4x.
